File: app/src/main/java/com/savvasdalkitsis/gameframe/draw/model/Palette.java
package com.savvasdalkitsis.gameframe.draw.model;

import com.savvasdalkitsis.gameframe.model.Moment;

import java.util.Arrays;

import lombok.Getter;
import lombok.experimental.Builder;

@Builder
@Getter
public class Palette implements Moment<Palette> {

    private String title;
    private boolean isSelected;
    private int[] colors;

    public void changeColor(int index, int color) {
        colors[index] = color;
    }

    public void setSelected(boolean selected) {
        this.isSelected = selected;
    }

    @Override
    public Palette replicateMoment() {
        int[] colorsCopy = copy(colors);
        return builder()
                .colors(colorsCopy)
                .isSelected(isSelected)
                .title(title)
                .build();
    }

    @Override
    public boolean isIdenticalTo(Palette moment) {
        return isSelected == moment.isSelected
                && title.equals(moment.title)
                && Arrays.equals(colors, moment.colors);
    }

    public static PaletteBuilder from(Palette palette) {
        return builder()
                .title(palette.title)
                .isSelected(palette.isSelected)
                .colors(copy(palette.colors));
    }

    private static int[] copy(int[] colors) {
        int[] colorsCopy = new int[colors.length];
        System.arraycopy(colors, 0, colorsCopy, 0,  colors.length);
        return colorsCopy;
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/draw/model/Palette.kt
package com.savvasdalkitsis.gameframe.feature.draw.model

import com.savvasdalkitsis.gameframe.feature.history.model.Moment
import java.util.*

data class Palette(
        var title: String,
        var isSelected: Boolean = false,
        val colors: IntArray) : Moment<Palette> {

    fun changeColor(index: Int, color: Int) {
        colors[index] = color
    }

    override fun replicateMoment() = this.copy(colors = colors.copyOf())

    override fun isIdenticalTo(moment: Palette) = this == moment

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Palette

        if (title != other.title) return false
        if (isSelected != other.isSelected) return false
        if (!Arrays.equals(colors, other.colors)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = title.hashCode()
        result = 31 * result + isSelected.hashCode()
        result = 31 * result + Arrays.hashCode(colors)
        return result
    }
}
